docs/enhancements/ipxe-host-boot.md

```diff
@@ -0,0 +1,125 @@
+assisted-service. This has been found to be at least as common a method 
+as providing the live ISO via virtual media and is also required in some 
+environments. Users are already unpacking the ISO we provide into PXE 
+artifacts and serving those in their infrastructure. We can make this much 
+easier for them by serving those artifacts as well as the iPXE boot script 
+directly from our application.
+
+### Goals
+
+- Serve iPXE artifacts (kernel, rootfs, and initrd)
+- Serve an iPXE boot script
+- Include the ignition in the provided initrd
+
+### Non-Goals
+
+- Support PXE (tftp based)
+- Integration with ironic (separate effort)
+- iPXE boot script customization (maybe in a follow-up if it’s useful)
+- Configure the minimal ISO to pull the rootfs from the image service
+
+## Proposal
+
+The image-service will serve the kernel, rootfs, and initrd with ignition 
+appended (see Implementation Details), while the assisted service will 
+provide an API to download the boot script. The boot script created and served 
+by the assisted service will contain URLs to the required artifacts hosted by 
+the image service.
+
+### User Stories
+
+#### Story 1
+
+As a user of assisted installer I want to boot my hosts using iPXE without 
+setting up my own http server to host artifacts.
+
+### Implementation Details/Notes/Constraints
+
+#### Ironic Integration
+
+To support integration with Ironic’s existing iPXE boot methods, the 
+discovery ignition must be appended to the initrd rather than specified as 
+a separate kernel parameter. The image service will also handle this initrd 
+manipulation when an initrd is requested for a particular infra-env.
+
+#### New Image Service APIs
+
+Image service APIs will be added to download the required boot artifacts. The 
+rootfs and kernel will be served directly from the template image. The initrd 
+endpoint will be specific to an infra-env in order to embed the ignition.
+
+- `GET /boot-artifacts/{artifact-name}?{params}`
+  - artifact-name can be either “rootfs” or “kernel”
+  - params must include “version” and “arch” which are used to identify a template image
+  - No authentication is required as we’re serving these directly from the uncustomized image
+  - Examples
+    - `/boot-artifacts/rootfs?version=4.9&arch=x86_64`
+    - `/boot-artifacts/kernel?version=4.10&arch=arm64`
+- `GET /images/{image-id}/pxe-initrd?{params}`
+  - image-id is the uuid of an infra-env in assisted-service
+  - params must include “version” and “arch”
+  - params may include “api_key” or “image_token” depending on assisted-service authentication configuration
+  - Example
+    - `/images/afe293b9-e3a0-47ed-8952-f26721626497/pxe-initrd?version=4.10&arch=x86_64&image_token={token}`
+
+#### New Assisted Service REST APIs
+
+Assisted service APIs will be added to download the ipxe boot script and to 
+retrieve a presigned url for downloading the boot script.
+
+- `GET /v2/infra-envs/{infra-env-id}/downloads/ipxe-script`
```

Comment:
You could convert all URL params to kernel params.

Reply:
That's interesting. That's assuming users are manipulating the original url we give them though which probably isn't the UX we're looking for (at least in the cloud).

FWIW we don't have a way to alter kernel parameters for the discovery iso today so it wouldn't be a regression if we just provide the same parameters that currently exist in the iso boot config.

Reply:
I think this may be a good direction for specifying kernel arguments in the future, but not really in the scope of this enhancement (in fact I specifically have "iPXE boot script customization" in the non-goals section already).